modules/katamari/src/main/ts/ephox/katamari/api/Optional.ts

```diff
@@ -1,141 +1,277 @@
-import * as Fun from './Fun';
+import * as Type from './Type';
 
-/** A value which may be Optional.some(a) or Optional.none().
- *  Useful for representing partiality or pass/fail conditions that produce a value when successful.
- *  Sometimes called a "type-safe alternative to null".
- *  Can be thought of as a list of exactly zero or 1 elements.
+/**
+ * The `Optional` type represents a value (of any type) that potentially does
+ * not exist. Any `Optional<T>` can either be a `Some<T>` (in which case the
+ * value does exist) or a `None` (in which case the value does not exist). This
+ * file defines a whole lot of FP-inspired utility functions for dealing with
+ * `Optional` objects. Note: Functions are grouped based on, primarily, their
+ * Haskell roots. That said, I'm not a Haskell programmer so it might be
+ * imperfect.
+ *
+ * Comparison with null or undefined
+ * - We don't get fancy null coalescing operators with `Optional`
+ * - We do get fancy helper functions with `Optional`
+ * - `Optional` support nesting, and allow for the type to still be nullable (or
+ * another `Optional`)
+ * - There is no option to turn off strict-optional-checks like there is for
+ * strict-null-checks
+ * - Try to use `Optional` instead of null or undefined where you can
  */
-export interface Optional<T> {
-  /** If none, run whenNone; if some(a) run whenSome(a) */
-  readonly fold: <U> (whenNone: () => U, whenSome: (v: T) => U) => U;
 
-  readonly isSome: () => boolean;
-  readonly isNone: () => boolean;
+export class Optional<T> {
+  private readonly tag: boolean;
```

Comment:
Can we use a string/enum here instead of a boolean? It'll keep it consistent with other discriminate unions we have and make it clearer when debugging.

Reply:
Looking more I can see this was done to be able to use truthy checks later on, so it's probably worth the payoff of using booleans.